### PR TITLE
UI: Topo Viz Tooltips

### PR DESCRIPTION
**Pull request by DingoEatingFuzz**
See some allocation information at a glance.

image

A lot of this diff is a CSS indentation change.

### CODE REVIEW

File: ui/tests/integration/components/topo-viz/node-test.js [hunk 1]
```diff
@@ -269,18 +276,72 @@ module('Integration | Component | TopoViz::Node', function(hooks) {
...
+
+    await TopoVizNode.memoryRects[0].hover();
+    assert.equal(this.onAllocationFocus.callCount, 1);
+    assert.equal(this.onAllocationBlur.callCount, 0);
+
+    await TopoVizNode.memoryRects[0].mouseleave();
+    assert.equal(this.onAllocationBlur.callCount, 0);
+
+    await TopoVizNode.mouseout();
+    assert.equal(this.onAllocationBlur.callCount, 1);
```
Comment (backspace) [inline ui/tests/integration/components/topo-viz/node-test.js:344]: Nice tests for mouse events 🤩

File: ui/tests/integration/components/topo-viz/node-test.js [hunk 2]
```diff
@@ -269,18 +276,72 @@ module('Integration | Component | TopoViz::Node', function(hooks) {
...
-    assert.ok(this.onAllocationSelect.callCount, 3);
+    assert.equal(this.onAllocationSelect.callCount, 3);
     assert.ok(this.onAllocationSelect.calledWith(this.node.allocations[1]));
 
     await TopoVizNode.memoryRects[1].select();
-    assert.ok(this.onAllocationSelect.callCount, 4);
+    assert.equal(this.onAllocationSelect.callCount, 4);
+  });
+
+  test('hovering over a memory or cpu rect for an alocation will call onAllocationFocus', async function(assert) {
```
Comment (backspace) [inline ui/tests/integration/components/topo-viz/node-test.js:293]: tiny typo: `alocation`

File: ui/app/utils/properties/style-string.js [hunk 3]
```diff
@@ -8,7 +8,7 @@ import { htmlSafe } from '@ember/template';
 //     styleStr: styleStringProperty('styleProps') // color:#FF0;border-width:1px
 export default function styleStringProperty(prop) {
   return computed(prop, function() {
-    const styles = this.get(prop);
+    const styles = get(this, prop);
```
Comment (backspace) [inline ui/app/utils/properties/style-string.js:11]: I briefly tried using this in the DAS realm and it failed, maybe fixing it was this simple??? haha oops 🙃

File: ui/app/components/topo-viz.js [hunk 4]
```diff
@@ -144,6 +153,19 @@ export default class TopoViz extends Component {
     if (this.args.onNodeSelect) this.args.onNodeSelect(this.activeNode);
   }
 
+  @action showTooltip(allocation, element) {
+    const bbox = element.getBoundingClientRect();
+    this.highlightAllocation = allocation;
+    this.tooltipProps = {
+      left: window.visualViewport.pageLeft + bbox.left + bbox.width / 2,
+      top: window.visualViewport.pageTop + bbox.top,
```
Comment (backspace) [inline ui/app/components/topo-viz.js:161]: I hadn’t heard of `visualViewport`! Sadly at the moment it’s feature flagged in Firefox, so it didn’t work for me until I tried it in Safari and Chrome 😞

image

**Review by backspace — APPROVED** [review]
It seems sad to me to not have Firefox support so I wonder if it’s possible to rework the calculations for that, but maybe it’s not worth it or something to revisit? Otherwise, this seems like a nice improvement 🎉

File: ui/app/components/topo-viz.js (hunk 4 above)
Comment (DingoEatingFuzz) [reply ui/app/components/topo-viz.js:161]: Curse the internet's betrayal of my trust. I'll swap this out with an older approach.

**Review by backspace — APPROVED** [review]
Total success! 😀

**Pull request merged**